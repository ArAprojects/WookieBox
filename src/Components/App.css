* {
  margin: 0;
  box-sizing: border-box;
  font-family: 'distant_galaxyregular';
}

body {
  background: url('/../swapi-background.jpg');
  background-attachment: fixed;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  overflow: auto;
}

@font-face {
  font-family: 'distant_galaxyregular';
  src: url('/../distgrg_-webfont.woff2') format('woff2'),
       url('/../distgrg_-webfont.woff') format('woff');
  font-weight: normal;
  font-style: normal;
}

header {
  text-align: center;
  /* font-family: 'distant_galaxyregular'; */
  color: yellow;
  font-size: 100px;
}

.buttonContainer {
display: flex;
justify-content: center;
flex-basis: auto;
margin: auto;
width: 100%;
margin-bottom: 25px;
}

.nav {
  display: flex;
  justify-content: center;
  align-content: center;
  flex-direction: column;
  margin-right: 100px;
  margin: auto;
  text-decoration: none;
}

input {
  height: 275px;
  width: 275px;
  padding: 0;
  border: none;
}

button {
  border: none;
  color: yellow;
  font-size: 45px;
  padding: 0;
  background-color: transparent;
  border: none;
}

#people {
  background: url(/../wookie.png) no-repeat;
}

#planets {
  background: url(/../death-star.png) no-repeat;
}

#vehicles {
  background: url(/../star-wars-vehicle.png) no-repeat;
}

#favorites {
  background: url(/../star.png) no-repeat;
}

.Loading {
  height: 200px;
  width: 200px;
  color: yellow;
}
