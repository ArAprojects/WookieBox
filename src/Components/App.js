import React, { Component } from 'react';
import Container from './Container.js'
import './App.css';
import {Route, NavLink} from 'react-router-dom';
import Card from './Card.js'

class App extends Component {
  constructor() {
    super();
    this.state = {
      people: [],
      planets: [],
      vehicles: [],
      crawlTextArray: [],
      favorites: []
    }
  }

  componentDidMount() {
  Promise.all([
    fetch("https://swapi.co/api/planets").then(response => response.json()).then(data => this.fetchEverything(data.results, ["residents"])),
    fetch("https://swapi.co/api/people").then(response => response.json()).then(data => this.fetchEverything(data.results, ["species", "homeworld"])),
    fetch("https://swapi.co/api/vehicles").then(response => response.json()).then(data => this.fetchEverything(data.results, ["films"])),
    fetch("https://swapi.co/api/films").then(response => response.json()).then(data => data.results).then(data => data.map(film => film.opening_crawl))
  ])
  .then(data => this.setState({people: data[1], planets: data[0], vehicles: data[2], crawlTextArray: data[3]}))
}


fetchEverything = (dataArray, keyArray) => {
const promises = dataArray.map(item => {
  let finalData = keyArray.map(keyName => {
    if (Array.isArray(item[keyName])) {
      let nestedData = item[keyName].map(nestedURL => fetch(nestedURL).then(response => response.json()))
      return Promise.all(nestedData)
    }
    else {
      return fetch(item[keyName]).then(response => response.json())
    }
  })
  return Promise.all(finalData)
   .then(data => ({data: item, nestedApis: data.flat()}))
   .catch(error => console.log(error));
});
return Promise.all(promises);
};

  render() {
    console.log(this.state)
    return (
      <div className="App">
        <header className="App-header">
          <h1>WookieBox</h1>
        </header>
        <section className="buttonContainer">
          <NavLink to='/people' className='nav'><button id="people">People</button></NavLink>
          <NavLink to='/planets' className='nav'><button id="planets">Planets</button></NavLink>
          <NavLink to='/vehicles' className='nav'><button id="vehicles">Vehicles</button></NavLink>
        </section>
        <Route path='/people' render = { () => <Card data = {this.state.people} />} />
        <Route path='/planets' render = { () => <Card data = {this.state.planets} />} />
        <Route path='/vehicles' render = { () => <Card data = {this.state.vehicles} />} />
      </div>

    );
  }

}

export default App;
