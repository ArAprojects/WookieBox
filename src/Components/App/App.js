import React, { Component } from 'react';
import './App.css';
import {Route, NavLink} from 'react-router-dom';
import Quote from '../Quote/Quote'
import Card from '../Card/Card'

class App extends Component {
  constructor() {
    super();
    this.state = {
      people: [],
      planets: [],
      vehicles: [],
      crawlText: [],
      favorites: []
    }
  }

  componentDidMount() {
    fetch("https://swapi.co/api/films").then(response => response.json()).then(data => data.results)
    .then(data => this.setState({crawlText: data[`${Math.floor(Math.random() * Math.floor(7))}`]}))
  Promise.all([
    fetch("https://swapi.co/api/planets").then(response => response.json()).then(data => this.fetchNested(data.results, ["residents"])),
    fetch("https://swapi.co/api/people").then(response => response.json()).then(data => this.fetchNested(data.results, ["species", "homeworld"])),
    fetch("https://swapi.co/api/vehicles").then(response => response.json()).then(data => data.results)
  ])
  .then(data => this.setState({people: data[1], planets: data[0], vehicles: data[2]}))
}

fetchNested = (dataArray, keyArray) => {
  const promises = dataArray.map(item => {
    let finalData = keyArray.map(keyName => {
      if (Array.isArray(item[keyName])) {
        let nestedData = item[keyName].map(nestedURL => fetch(nestedURL).then(response => response.json()))
        return Promise.all(nestedData)
      } else {
        return fetch(item[keyName]).then(response => response.json())
      }
    })
  return Promise.all(finalData)
    .then(data => ({...item, [keyArray[0]]: data[0].flat(), [keyArray[1]]: data[1]}))
    .catch(error => console.log(error));
  });
  return Promise.all(promises);
};


toggleFavorite = (favorite) => {
  if(!favorite.favorite) {
    favorite["favorite"] = true;
    this.setState({ favorites: [...this.state.favorites, favorite] });
  } else {
    let index = this.state.favorites.indexOf(favorite);
    this.state.favorites.splice(index, 1);
    this.setState({...this.state.favorites})
    favorite["favorite"] = false;
  }

}

  render() {
    return (
      <div className="App">
        <header className="App-header">
          <h1>WookieBox</h1>
        </header>
        <section className="buttonContainer">
          <NavLink to='/people' className='nav'><input type="submit" id="people" value=""/><button type="submit">People</button></NavLink>
          <NavLink to='/planets' className='nav'><input type="submit" id="planets" value=""/><button type="submit">Planets</button></NavLink>
          <NavLink to='/vehicles' className='nav'><input type="submit" id="vehicles" value=""/><button type="submit">Vehicles</button></NavLink>
          <NavLink to='/favorites' className='nav'><input type='submit' id="favorites" values=""/><button type="submit">Favorites({this.state.favorites.length})</button></NavLink>
        </section>
        <section>
          <Route exact path="/" render = {() => <Quote className="Quote" data = {this.state.crawlText}/>}/>
          <Route path='/people' render = { () => <Card className="Card" data = {this.state.people} toggleFavorite = {this.toggleFavorite} />} />
          <Route path='/planets' render = { () => <Card className="Card" data = {this.state.planets} toggleFavorite = {this.toggleFavorite} favorites = {this.state.favorites} />} />
          <Route path='/vehicles' render = { () => <Card className="Card" data = {this.state.vehicles} toggleFavorite = {this.toggleFavorite} />} />
          <Route path='/favorites' render = { () => <Card className="Card" data = {this.state.favorites} toggleFavorite = {this.toggleFavorite}/>} />
        </section>
      </div>

    );
  }

}

export default App;
