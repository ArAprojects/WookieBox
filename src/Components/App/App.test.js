import React from 'react';
import { shallow } from 'enzyme';
import App from './App';

describe('App', () => {
  it('works', () => {
    const wrapper = shallow(<App />)
    expect(wrapper).toMatchSnapshot()
  })
});
