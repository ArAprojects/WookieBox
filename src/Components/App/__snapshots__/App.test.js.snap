// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App works 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <h1>
      WookieBox
    </h1>
  </header>
  <section
    className="buttonContainer"
  >
    <NavLink
      className="nav"
      to="/people"
    >
      <input
        id="people"
        type="submit"
        value=""
      />
      <button
        type="submit"
      >
        People
      </button>
    </NavLink>
    <NavLink
      className="nav"
      to="/planets"
    >
      <input
        id="planets"
        type="submit"
        value=""
      />
      <button
        type="submit"
      >
        Planets
      </button>
    </NavLink>
    <NavLink
      className="nav"
      to="/vehicles"
    >
      <input
        id="vehicles"
        type="submit"
        value=""
      />
      <button
        type="submit"
      >
        Vehicles
      </button>
    </NavLink>
    <NavLink
      className="nav"
      to="/favorites"
    >
      <input
        id="favorites"
        type="submit"
        values=""
      />
      <button
        type="submit"
      >
        Favorites(
        0
        )
      </button>
    </NavLink>
  </section>
  <section>
    <Route
      exact={true}
      path="/"
      render={[Function]}
    />
    <Route
      path="/people"
      render={[Function]}
    />
    <Route
      path="/planets"
      render={[Function]}
    />
    <Route
      path="/vehicles"
      render={[Function]}
    />
    <Route
      path="/favorites"
      render={[Function]}
    />
  </section>
</div>
`;
